Крестики-нолики

Lbisiy opened the pull request:
Добавлены файлы main_new и logic_new
Пока сырой вариант: 1. То работает , то не работает проверка при введении ячейки больше, чем размерность поля. 2. Как реализовать проверку, если ячейка уже занята. Где это лучше делать внутри функции (какой?) или в основной программе? 3. Вроде игра пока заканчивается, если кто-то выиграл. 4. Как сделать проверку, закончилось ли поле. Где это лучше сделать? Внутри функции(какой?) или внутри самой программы? P.S. Очень тяжело идет, поэтому просьба указать на моменты, где неправильная логика и где надо переделать. Т.к. если поправите за меня, в голове у меня ничего не отложится. Заранее спасибо!

Lbisiy committed:
Крестики-нолики

Добавлены файлы main_new и logic_new
Пока сырой вариант: 1. То работает , то не работает проверка при введении ячейки больше, чем размерность поля. 2. Как реализовать проверку, если ячейка уже занята. Где это лучше делать внутри функции (какой?) или в основной программе? 3. Вроде игра пока заканчивается, если кто-то выиграл. 4. Как сделать проверку, закончилось ли поле. Где это лучше сделать? Внутри функции(какой?) или внутри самой программы? P.S. Очень тяжело идет, поэтому просьба указать на моменты, где неправильная логика и где надо переделать. Т.к. если поправите за меня, в голове у меня ничего не отложится. Заранее спасибо!

EgorOb commented on logic_new.py:
```diff
@@ -0,0 +1,87 @@
+            field[move] = 'X'
+            return field
+
+def player2_step(field, move) -> list:
```
player1_step и player2_step идентичные функции, отличающиеся только одним символом. Может тогда написать одну функцию, а отличающийся символ передать как входной параметр?

Lbisiy replied:
Я специально оставил две функции, чтобы потом одну заменить со временем на  random, чтобы играть с компьютером. Но пока еще не победил до конца игру 2-х игроков (

EgorOb commented on main_new.py:
```diff
@@ -0,0 +1,38 @@
+
+while True:
+    if PLAYER_1 == True:
+        while True:
```
Я бы посоветовал вынести эту проверку в отдельное место и назвать допустим так
```
def check_move(player):
    while True:
        move = int(input(f'Чтобы поставить {player} укажите номер ячейки:'))
        if move > pow(INDEX, 2) or move < 0:
            print('Вы ввели неправильный номер ячейки')
            continue
        else:
            break
    return move
```
В таком случае вы сможете ещё туда дописать проверок каких захотите

EgorOb commented on main_new.py:
```diff
@@ -0,0 +1,38 @@
+field = [' '] * pow(INDEX, 2)  # инициализируем пустое поле
+
+draw_field(field)  # рисуем пустое поле
+
```
Я бы посоветовал определить игрока до начала while, тогда можно в цикле реализовать автоматическую смену игрока в конце хода таким образом 
`current_player = "X" if current_player == "0" else "0"`

EgorOb replied:
Тогда ваш код можно будет уменьшить до шаблона, который позволит реализовать алгоритм игры
```
current_player = ...
while True:
    move = check_move(current_player)
    field = player_step(field, move, current_player)
    draw_field(field)
    win = who_win(field)
    continue_ = game_continue(win, move)
    if win != 0 or continue_ != 0:
        break
    current_player = "X" if current_player == "0" else "0"
```

EgorOb reviewed (commented):
Постарался не писать за вас, но менять конструкции используя ваш код. Попробуйте поправить код применяя концепции, что я показал. У вас много дублирующего кода, нужно стараться автоматизировать действия и быть немного ленивым, чтобы не писать вручную однотипные конструкции, а лучше их постараться объединить в функцию с изменяющимися переменными.

Lbisiy committed:
Крестики-нолики с правками

Добавлены файлы logic_new.py, main_new.py с изменениями. Пока не реализована проверка при ничьей. Если ничья пока продолжает предлагать вводить ячейку, также не отлавливает ошибку, если ничего не вводить в номер поля или вводить не цифру.

Lbisiy committed:
Крестики-нолики измененный logic.new.py

Уменьшил код в функции def draw_field()

Lbisiy committed:
Крестики-нолики

Добавлен файл game.py. Сделаны почти все функции и проверки. Осталось допилить функцию draw_field, пока она реализована для поля 3х3 только и по колхозному и внести проверку по ничьей. Ну и может еще добавить некоторые ошибки через трай, эксепт.

Lbisiy committed:
Крестики-нолики после всех доработок

Добавлен файл game.py. Вроде как реализован весь необходимый функционал

EgorOb reviewed (approved):
Вроде по коду нормально, описание есть, ничего не дублирует. Проект зачтён. Зачёт поставил
